### DISCUSSION TITLE
新建/编辑时隐藏某些列, 适应不同场景

### DISCUSSION THREAD

**Discussion by densen2014** [dotnetcore/BootstrapBlazor#2937] [Q&A] [answered]
### 说明

讨论添加特性支持新建时此列隐藏和编辑时此列隐藏, 适应不同场景. 节省制作新建模板和新建数据Dto操作, 入门使用更简单

- IsVisibleWhenAdd 新建时此列显示
- IsVisibleWhenEdit 编辑时此列显示
- Visibility 可见性属性枚举类型,  Unset 未设置 | Visible 可见 | Hidden 隐藏
- 如果有列设置了 VisibleWhenAdd 或者 VisibleWhenEdit 属性为 Visible, 隐藏其他列, 用于新建数据Dto场景

### 示例:

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/15415ae6-1cb9-464d-a001-451ce1856f66)

### 添加界面:

只显示Visible的单元格

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/cc006d35-18b5-4517-ba2c-2069697d0073)

### 编辑界面:

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/0edd2665-a13a-49d4-bda2-e44271728377)

### 代码

```
    <Table TItem="Foo" Items="@Items.Take(3)"  IsMultipleSelect="true"
           ShowToolbar="true" ShowExtendButtons="true" >
        <TableColumns>
            <TableColumn @bind-Field="@context.DateTime" Width="180" />
            <TableColumn @bind-Field="@context.Name" IsVisibleWhenAdd="Visibility.Visible" />
            <TableColumn @bind-Field="@context.Address" IsVisibleWhenEdit="Visibility.Hidden" />
            <TableColumn @bind-Field="@context.Count" IsVisibleWhenAdd="Visibility.Visible"/>
            <TableColumn @bind-Field="@context.Complete" IsVisibleWhenAdd="Visibility.Visible" />
            <TableColumn @bind-Field="@context.Education" IsVisibleWhenAdd="Visibility.Unset" />
        </TableColumns>
    </Table>
```

**Comment by densen2014** [comment]
### 特性示例

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/4e06bab4-17b2-4ec8-aa31-31e4a44e9ebb)

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/f32f4725-e259-487b-8e75-e0d3797bb1ca)


```
<Table TItem="Foo" Items="@Items.Take(3)"  IsMultipleSelect="true"
       ShowToolbar="true" ShowExtendButtons="true" AutoGenerateColumns > 
</Table>


public class Foo
{ 
    /// <summary>
    /// 主键
    /// </summary>
    [Key]
    [Display(Name = "主键")]
    [AutoGenerateColumn(Ignore = true)]
    public int Id { get; set; }

    /// <summary>
    /// 姓名
    /// </summary>
    [Required(ErrorMessage = "{0}不能为空")]
    [AutoGenerateColumn(Order = 10, Filterable = true, Searchable = true, IsVisibleWhenAdd = Visibility.Visible)]
    [Display(Name = "姓名")]
    public string? Name { get; set; }

    /// <summary>
    /// 日期
    /// </summary>
    [AutoGenerateColumn(Order = 1, FormatString = "yyyy-MM-dd", Width = 180)]
    [Display(Name = "日期")]
    public DateTime? DateTime { get; set; }

    /// <summary>
    /// 地址
    /// </summary>
    [Display(Name = "地址")]
    [Required(ErrorMessage = "{0}不能为空")]
    [AutoGenerateColumn(Order = 20, Filterable = true, Searchable = true, IsVisibleWhenAdd = Visibility.Hidden)]
    public string? Address { get; set; }

    /// <summary>
    /// 数量
    /// </summary>
    [Display(Name = "数量")]
    [Required]
    [AutoGenerateColumn(Order = 40, Sortable = true)]
    public int Count { get; set; }

    /// <summary>
    /// 是/否
    /// </summary>
    [Display(Name = "是/否")]
    [AutoGenerateColumn(Order = 50, IsVisibleWhenAdd = Visibility.Visible)]
    public bool Complete { get; set; }

    /// <summary>
    /// 学历
    /// </summary>
    [Required(ErrorMessage = "请选择学历")]
    [Display(Name = "学历")]
    [AutoGenerateColumn(Order = 60)]
    public EnumEducation? Education { get; set; }

    /// <summary>
    /// 爱好
    /// </summary>
    [Required(ErrorMessage = "请选择一种{0}")]
    [Display(Name = "爱好")]
    [AutoGenerateColumn(Order = 70, Editable = false)]
    public IEnumerable<string> Hobby { get; set; } = new List<string>();
}
```

**Comment by FelixYe8818** [comment]
是否还可以考虑增加一个IsLoad的特性？

目前我们的TableColum虽然已经有了一个Visible，但是如果ShowColumnList="true"，依旧会出现在ColumnList里面，目前我是使用一个IF来包TableColum来实现这个功能。

我感觉可以这样设计：IsLoad默认等于True，如果IsLoad等于False，那么在新建，编辑，显示，以及ColumnList里面都隐藏起来。

**Reply by FelixYe8818** [reply]
![11](https://github.com/dotnetcore/BootstrapBlazor/assets/148768346/a4cb78cf-c23b-4a99-8f8b-804ca5ab4723)

**Reply by densen2014** [reply]
```
public partial class Users
{

    public string? Id { get; set; }

    [DisplayName("用户名")]
    public string? UserName { get; set; }

    public string? Email { get; set; }

    public string? Password { get; set; }

}

public partial class AddUserDto : Users
{
    [AutoGenerateColumn(Ignore = true)] 
    new public string? Password { get; set; }

}
```

 用 AddUserDto  去渲染就忽略你那个列了

**Reply by ArgoZhang** [reply]
`Visible` 是与 `ShowColumnList` 联动的，就是这样设计的，`Visible` 就是 `ShowColumnList` 的初始值，永久不显示就是 `Ignoer`

**Reply by FelixYe8818** [reply]
我之前有一个场景，用户可以在后台设置：是否永久不显示某一列（我使用IF包TableColum），是否隐藏某一列（使用Visible控制），使用  [AutoGenerateColumn(Ignore = true)] 来控制的话，又无法动态的控制 Ignore 等于true还是等于false，只能一次性写死。

**Reply by ArgoZhang** [reply]
```
<TableColumn Ignore="IsIgnore" @bind-Field="@context.Name"></TableColumn>

@code {
    public bool IsIgnore { get; set; }
}
```

**Reply by FelixYe8818** [reply]
![11](https://github.com/dotnetcore/BootstrapBlazor/assets/148768346/8374bc31-9dda-4dcb-859b-33f09347d1ed)

我添加IsIgnore后好像没有生效。使用的是8.2.2版本

**Reply by ArgoZhang** [reply]
@FelixYe8818 Mark 稍后给出解决方案 `Ignore` 这里不符合语义，我想增加一个 `Show` 但是又与 `Visible` 有点混淆。我看看源码再设计

**Comment by ArgoZhang** [comment, marked as answer]
@densen2014 @FelixYe8818 

更新 `Table` 组件使用其接口 `Editable` 参数来处理这个需求

`Editable="false"` 时不渲染当前列，并不会出现在 `ColumnList` 中
`Visible="false"` 时不渲染当前列，并出现在 `ColumnList` 中

老版本中 `Editable` 参数未使用 `v8.4.1-beta02` 中更新此功能

相关示例 https://www.blazor.zone/table/column#Ignore

请点击下方 `Mark as answer`
